product/specs/2022-01-views/03-modeling-view-columns.md

```diff
@@ -0,0 +1,28 @@
+---
+title: Views: Modeling View Columns
+description: 
+published: true
+date: 2022-01-13T19:49:54Z
+tags: 
+editor: markdown
+dateCreated: 2022-01-13T19:49:54Z
+---
+
+Here's how I think we should model view columns in our API and UI. Each heading represents an attribute of a View Column.
+
+### Data Type
+- **Definition**: This is the final data type of the content of the column after any computations etc. are applied.
+- **Allowed values**: same as Table data types.
+- **Required**. Data type should always be set, at the very least, we can treat unknown data types as text.
+
+### Data Sources
+ - **Definition**: This is the set of source columns that are used to generate the data in the current View column.
+- **Allowed values**: references to other Table or View columns, including other columns in the same View.
+- **Optional**: This could be empty for purely calculated columns (e.g. using the Postgres `random()` function and putting the output in a column)
+
+### Data Formula
+- **Definition**: This is the formula used to generate data in for this column.
+- **Allowed values**: SQL function + operators
+- **Optional**: Columns that are direct copies of other columns from tables or views won't have a formula.
```

Comment:
Should Data sources and Data formula column types be specific to views?

I assume tables would also need them for generated columns created using GENERATED ALWAYS AS, and for auto-incremented columns.

For tables, we'd allow editing all columns if they are not a primary-key and it would fail if the connected db has tables with such columns. We should atleast have a way to identify them for tables to disable the input on the frontend.

Reply:
I think it would make sense to add Data Sources and Data Formula to generated columns in tables, I hadn't thought of that.

Reply:
Added an issue to track product work for this: https://github.com/centerofci/mathesar/issues/1002